### PR TITLE
stats: adds metrics sink with ack

### PR DESCRIPTION
**Pull request by jingwei99**
Description: Replace the metrics sink from the envoy one to a custom one. The custom metrics sink comes with ack function on the grpc stream. This change is based on upstream envoy change: https://github.com/envoyproxy/envoy/pull/13919
Risk Level: High
Testing: Local and ci unit tests and integration tests

Co-authored-by: Jose Nino jnino@lyft.com
Signed-off-by: Jingwei Hao <jingweih@lyft.com>

### CODE REVIEW

File: library/common/extensions/stat_sinks/metrics_service/service.proto [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+  }
+
+  // Identifier data effectively is a structured metadata. As a performance optimization this will
+  // only be sent in the first message on the stream.
+  Identifier identifier = 1;
+
+  // A list of metric entries
+  repeated io.prometheus.client.MetricFamily envoy_metrics = 2;
+
+  string batch_id = 3;
```
Comment (junr03) [inline library/common/extensions/stat_sinks/metrics_service/service.proto:33]: nit: I would make this field 2.

Also lets document what this field is for.

File: library/common/config_template.cc [hunk 2]
```diff
@@ -207,6 +207,13 @@ stats_flush_interval: {{ stats_flush_interval_seconds }}s
       grpc_service:
         envoy_grpc:
           cluster_name: stats
+  - name: envoy.stat_sinks.metrics_service.mobile
```
Comment (junr03) [inline library/common/config_template.cc:210]: I am going to put up a PR today to reduce some of the stats we are emitting (and no longer need), as this config effectively doubles the amount of stats emitted

**Review by junr03 — COMMENTED** [review]
mostly lgtm! Awesome. A small comment

File: library/common/extensions/stat_sinks/metrics_service/service.proto (hunk 1 above)
Comment (jingwei99) [reply library/common/extensions/stat_sinks/metrics_service/service.proto:33]: you mean 
```
string batch_id = 2;
repeated io.prometheus.client.MetricFamily envoy_metrics = 3;
```
?

Comment (junr03) [reply library/common/extensions/stat_sinks/metrics_service/service.proto:33]: Yes

File: library/common/extensions/stat_sinks/metrics_service/service.proto [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+import "validate/validate.proto";
+
+service EnvoyMobileMetricsService {
+  rpc EnvoyMobileStreamMetrics(stream EnvoyMobileStreamMetricsMessage)
+      returns (stream EnvoyMobileStreamMetricsResponse) {
+  }
+}
+
+message EnvoyMobileStreamMetricsResponse {
+  string batch_id = 1;
```
Comment (junr03) [inline library/common/extensions/stat_sinks/metrics_service/service.proto:16]: comment

File: library/common/extensions/stat_sinks/metrics_service/service.proto [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+  message Identifier {
+    // The node sending metrics over the stream.
+    envoy.config.core.v3.Node node = 1 [(validate.rules).message = {required: true}];
+  }
+
+  // Identifier data effectively is a structured metadata. As a performance optimization this will
+  // only be sent in the first message on the stream.
+  Identifier identifier = 1;
+
+  // Identifier for the current envoy_metrics batch
```
Comment (junr03) [inline library/common/extensions/stat_sinks/metrics_service/service.proto:30]:
```suggestion
  // Identifier for the current envoy_metrics batch.
```

Same below.

**Review by junr03 — APPROVED** [review]

**Pull request merged**